### PR TITLE
Model callback refactor and add undo support for delete to first post only

### PR DESCRIPTION
**Pull request by angusmcleod**
@pmusaraj Apologies but this includes two things in one. I've tried to keep the PRs discrete, but I have a few backed up. As such I'm going to give you a bit more narrative here.

This is mostly a refactor of what was handle_model_callback in DiscourseActivityPubObject. That method, and its companion method clear_all_objects, have been moved to Concerns to make them more discrete and clearer. All the tests are the same (with some additional), just moved into new files.

In the new model callback system you'll also notice some new "hooks" such as ``before_perform_activity_pub_activity``. That is part of a larger theme of progressively making the separation of concerns as clear as possible, i.e.

> AP (Pure ActivityPub) -> DiscourseActivityPub (The Plugin) -> plugin.rb (where the plugin integrates with Discourse)

You'll see this develop further in subsequent PRs.

This also adds "Undo" support for delete (i.e. supporting Discourse's post restore), which was the original intent of this branch, but blink and you'll miss it. It's this line

```
on(:post_recovered) do |post, opts, user|
   post.perform_activity_pub_activity(:create)
end
```

Basically, Mastodon expects an entirely new Create instead of an Undo of a previous Delete.

### CODE REVIEW

**Review by pmusaraj — APPROVED** [review]
Thanks for the explanation. I will merge shortly, I only have a question about a test for the recovered post use case (but that can be addressed later as well).

File: plugin.rb [hunk 1]
```diff
@@ -344,12 +338,15 @@
...
   on(:post_created) do |post, opts, user|
-    DiscourseActivityPubObject.handle_model_callback(post, :create)
+    post.perform_activity_pub_activity(:create)
   end
   on(:post_destroyed) do |post, opts, user|
-    DiscourseActivityPubObject.handle_model_callback(post, :delete)
+    post.perform_activity_pub_activity(:delete)
+  end
+  on(:post_recovered) do |post, opts, user|
+    post.perform_activity_pub_activity(:create)
```
[the remark this answers is not in the document]
Comment (angusmcleod) [reply plugin.rb:350]: What stopped me from adding a test is the thought I would just be testing that `on(:post_recovered)` works. The create action is covered in tests. I felt it would be a false positive. But maybe I was wrong.

[the remark this answers is not in the document]
Comment (pmusaraj) [reply plugin.rb:350]: Yeah, true. A test would only ensure that someone in the future doesn't inadvertently delete the `on(:post_recovered)` lines.
